Introduce Translatable inputs and fields (#5810)
marmelab/react-admin: docs/Inputs.md

Check [the `ra-relationships` documentation](https://marmelab.com/ra-enterprise/modules/ra-relationships#referencemanytomanyinput) for more details.

## Translatable Inputs

> You may have inputs which are translated in multiple languages and want users to edit translations for each language separately. To display them, you can use the `<TranslatableInputs>` components which expect the translatable fields to have the following structure:

Reviewer: ```suggestion
You may have inputs which are translated in multiple languages and want users to edit translations for each language separately. To display them, you can use the `<TranslatableInputs>` component, which expects the translatable values to have the following structure:
```

```js
{
    name: {
        en: 'The english value',
        fr: 'The french value',
        tlh: 'The klingon value',
    }

> }

Reviewer: same, please add a value for the description field

Author: Done

```

This is how to use it:

```jsx
<TranslatableInputs locales={['en', 'fr']}>
    <TextInput source="name">
    <RichTextInput source="description">
</TranslatableInputs>
```

> 

Reviewer: I think you should have a paragraph about validation, explaining that errors in any of the locales will highlight the tab header as invalid.

> You may override the locale selected by default using the `defaultLocale` prop which react-admin sets to the current UI locale when not provided.

Reviewer: ```suggestion
React-admin uses the user locale as the default locale in this field. You can override this setting by default using the `defaultLocale` prop:
```

```jsx
<TranslatableInputs locales={['en', 'fr']} defaultLocale="fr">
    <TextInput source="name">
    <RichTextInput source="description">
</TranslatableInputs>
```

By default, `<TranslatableInputs>` will allow users to select the displayed locale using Material-ui tabs with the locale code as their labels.

> You may override the tabs labels using translation keys following this format: 'ra.locales.locale_code'. For example `ra.locales.en` or `ra.locales.fr`.

Reviewer: ```suggestion
You may override the tabs labels using translation keys following this format: 'ra.locales.[locale_code]'. For instance, `ra.locales.en` or `ra.locales.fr`.
```

> You may override the language selector using the `selector` prop which accept a React element:

Reviewer: ```suggestion
You may override the language selector using the `selector` prop, which accepts a React element:
```

```jsx
const Selector = () => {
    const {
        locales,
        selectLocale,
        selectedLocale,
    } = useTranslatableContext();

    const handleChange = (event): void => {
        selectLocale(event.target.value);
    };

    return (
        <select
            aria-label="Select the locale"
            onChange={handleChange}
            value={selectedLocale}
        >
            {locales.map(locale => (
                <option
                    key={locale}
                    value={locale}

>                     // This allows to correctly links the containers for each locale to their labels

Reviewer: ```suggestion
                    // This allows to correctly link the containers for each locale to their labels
```

                    id={`translatable-header-${locale}`}
                >
                    {locale}
                </option>
            ))}
        </select>
    );
};

<TranslatableInputs
    record={record}
    resource="products"
    basePath="/products"
    locales={['en', 'fr']}
    selector={<Selector />}
>
    <TextInput source="name" />
    <RichTextInput source="description" />
</TranslatableInputs>
```

If you have multiple `TranslatableInputs` on the same page, you should specify a `groupKey` so that react-admin can create unique identifiers for accessibility.

```jsx
<TranslatableInputs locales={['en', 'fr']} groupKey="essential-fields">
    <TextInput source="name">
    <RichTextInput source="description">
</TranslatableInputs>
```

## Recipes 

> ### Transforming Input Value to/from Record

Reviewer: I think you should have a section about TranslatableFields and Inputs in the i18n documentation as well.

Author: You mean a very simple one redirecting to the fields and inputs pages?

Reviewer: absolutely!
